### ISSUE TITLE
Possible to combine It -TestCases and Mock When Mocking Cmdlet from Imported Module?

### ISSUE THREAD

**Issue by mhendric** [pester/Pester#1086]
#### Issue

I wanted to see if it’s currently possible to Mock a cmdlet from an imported module, and also use the -TestCases parameter for the It block? While I’m able to combine TestCases and Mock when mocking a cmdlet from the default scope, it appears that Mock does not work at all when I’m using TestCases, and the mocked cmdlet is from an imported module. When that is the case, the function that is supposed to be Mocked doesn’t get overridden at all. Is this just a syntax problem on my part, or is this something Pester is not actually currently capable of?

Here’s a simple test script and test module which demonstrates this problem:

#### Test Output:
```
PS D:\> Invoke-Pester .\MockTest.ps1
Executing all tests in '.\MockTest.ps1'

Executing script .\MockTest.ps1

  Describing Pester Mock and TestCases Test
    [+] Without Test Cases 1.11s
    [+] Without Test Cases Different Module 172ms
    [+] With Test Cases 94ms
    [-] With Test Cases Different Module - Returning Param 90ms
      Expected 'ec98deb6-0bd6-4573-b819-5e51c33426ae', but got dae15760-f6ea-4fe7-94cf-69fe1715ff53.
      41:         New-Guid2 | Should be $Guid
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, D:\MockTest.ps1: line 41
    [-] With Test Cases Different Module - Not Returning Param 124ms
      Expected <empty>, but got 3228b1c7-2811-4229-89e1-ea63463bc956.
      51:         New-Guid2 | Should be ""
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, D:\MockTest.ps1: line 51
    [-] With Test Cases Different Module - No Param 30ms
      Expected <empty>, but got c9f07703-be8f-43e4-b377-4fb0e646fca7.
      59:         New-Guid2 | Should be ""
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, D:\MockTest.ps1: line 59
Tests completed in 1.62s
Tests Passed: 3, Failed: 3, Skipped: 0, Pending: 0, Inconclusive: 0 
```

#### MockTest.ps1
```
Get-Module TestModule | Remove-Module
Import-Module .\TestModule.psm1

Describe 'Pester Mock and TestCases Test' {
    $testGuid = "ec98deb6-0bd6-4573-b819-5e51c33426ae"
    $testCases = @(@{Guid=$testGuid})

    It 'Without Test Cases' {
        Mock -CommandName New-Guid -MockWith {
            return $testGuid
        }

        [string]::IsNullOrEmpty((New-Guid))| Should be $false
    }

    It 'Without Test Cases Different Module' {
        Mock -CommandName New-Guid2 -ModuleName TestModule -MockWith {
            return $testGuid
        }

        [string]::IsNullOrEmpty((New-Guid2))| Should be $false
    }

    It 'With Test Cases' -TestCases $testCases {
        param($Guid)

        Mock -CommandName New-Guid -MockWith {
            return $Guid
        }

        New-Guid | Should be $Guid
    }

    It 'With Test Cases Different Module - Returning Param' -TestCases $testCases {
        param($Guid)

        Mock -CommandName New-Guid2 -ModuleName TestModule -MockWith {
            return $Guid
        }

        New-Guid2 | Should be $Guid
    }

    It 'With Test Cases Different Module - Not Returning Param' -TestCases $testCases {
        param($Guid)

        Mock -CommandName New-Guid2 -ModuleName TestModule -MockWith {
            return ""
        }

        New-Guid2 | Should be ""
    }

    It 'With Test Cases Different Module - No Param' -TestCases $testCases {
        Mock -CommandName New-Guid2 -ModuleName TestModule -MockWith {
            return ""
        }

        New-Guid2 | Should be ""
    }
}
```

#### TestModule.psm1
```
function New-Guid2
{
    return (New-Guid)
}

Export-ModuleMember *
```

#### Environment

Windows 10 build 1803 - 17134.165
PowerShell Version: 5.1.17134.165
Pester Version: 4.4.0

**Comment by nohwnd** [comment]
I am getting different results, and also different results based on how many times I run the tests. I need to test this on Windows, but I only have my mac on me. :/

**Comment by mhendric** [comment]
If you're getting failures because of a lack of the New-Guid cmdlet (which happens to me on my Windows 2012 R2 / PowerShell 4 machines), I can also reproduce this by replacing every instance of New-Guid with Get-Process (including changing New-Guid2 to Get-Process2).

**Comment by johlju** [comment]
You are calling the same cmdlet as you are mocking. If you are mocking `New-Guid` instead of `New-Guid2`, then the tests are passing. 

```powershell
Get-Module TestModule | Remove-Module
Import-Module .\TestModule.psm1

Describe 'Pester Mock and TestCases Test' {
    $testGuid = "ec98deb6-0bd6-4573-b819-5e51c33426ae"
    $testCases = @(@{Guid=$testGuid})

    It 'Without Test Cases' {
        Mock -CommandName New-Guid -MockWith {
            return $testGuid
        }

        [string]::IsNullOrEmpty((New-Guid))| Should be $false
    }

    It 'Without Test Cases Different Module' {
        Mock -CommandName New-Guid2 -ModuleName TestModule -MockWith {
            return $testGuid
        }

        [string]::IsNullOrEmpty((New-Guid2))| Should be $false
    }

    It 'With Test Cases' -TestCases $testCases {
        param($Guid)

        Mock -CommandName New-Guid -MockWith {
            return $Guid
        }

        New-Guid | Should be $Guid
    }

    It 'With Test Cases Different Module - Returning Param' -TestCases $testCases {
        param($Guid)

        Mock -CommandName New-Guid -ModuleName TestModule -MockWith {
            return $Guid
        }

        New-Guid2 | Should be $Guid
    }

    It 'With Test Cases Different Module - Not Returning Param' -TestCases $testCases {
        param($Guid)

        Mock -CommandName New-Guid -ModuleName TestModule -MockWith {
            return ""
        }

        New-Guid2 | Should be ""
    }

    It 'With Test Cases Different Module - No Param' -TestCases $testCases {
        Mock -CommandName New-Guid -ModuleName TestModule -MockWith {
            return ""
        }

        New-Guid2 | Should be ""
    }
}
```

**Comment by johlju** [comment]
Hmm.. .They are passing in VSCode, but not in the console 🤔 

```
Describing Pester Mock and TestCases Test
  [+] Without Test Cases 18ms
  [+] Without Test Cases Different Module 14ms
  [+] With Test Cases 10ms
  [+] With Test Cases Different Module - Returning Param 12ms
  [+] With Test Cases Different Module - Not Returning Param 12ms
  [+] With Test Cases Different Module - No Param 12ms
```

But in the console one is failing
```
Executing all tests in '.\MockTest.Tests.ps1'

Executing script .\MockTest.Tests.ps1

  Describing Pester Mock and TestCases Test
    [+] Without Test Cases 53ms
    [+] Without Test Cases Different Module 12ms
    [+] With Test Cases 11ms
    [-] With Test Cases Different Module - Returning Param 15ms
      Expected 'ec98deb6-0bd6-4573-b819-5e51c33426ae', but got $null.
      41:         New-Guid2 | Should be $Guid
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\Pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, C:\Source\HelpUsers\mike\guidtests\MockTest.Tests.ps1: line 41
    [+] With Test Cases Different Module - Not Returning Param 17ms
    [+] With Test Cases Different Module - No Param 12ms
```

**Comment by mhendric** [comment]
My intention in this repro was in fact to Mock New-Guid2 (a function not in the default scope). For my test (of which this is just a simplified example), I want to override a function that is in another module, but the function doesn't get overridden when TestCases are used.

**Comment by johlju** [comment]
Ah sorry. Removing my comments to not pollute the thread. 🙂

**Comment by mhendric** [comment]
Not sure if it's the same issue or not, but I can also reproduce pretty much the same behavior using a foreach loop instead of the It -TestCases parameter. This uses the same TestModule.psm1 from above:

### MockForeachTest.ps1
```
Import-Module .\TestModule.psm1

Describe "Nested Module Test" {

    $testCases = @(
        @{TestName="Test1";String="aaa"},
        @{TestName="Test2";String="bbb"},
        @{TestName="Test3";String="ccc"}
    )

    foreach ($testCase in $testCases)
    {
        It "Same Scope Cmdlet: $($testCase.TestName)" {
            Mock -CommandName New-Guid -MockWith {return $testCase.String}
            New-Guid | Should be $testCase.String
        }
    }

    foreach ($testCase in $testCases)
    {
        It "Imported Cmdlet: $($testCase.TestName)" {
            Mock -CommandName New-Guid2 -ModuleName TestModule -MockWith {return $testCase.String}
            New-Guid2 | Should be $testCase.String
        }
    }
}
```

### Test Results
```

PS D:\> Invoke-Pester .\MockForeachTest.ps1
Executing all tests in '.\MockForeachTest.ps1'

Executing script .\MockForeachTest.ps1

  Describing Nested Module Test
    [+] Same Scope Cmdlet: Test1 1.16s
    [+] Same Scope Cmdlet: Test2 149ms
    [+] Same Scope Cmdlet: Test3 32ms
    [-] Imported Cmdlet: Test1 185ms
      Expected 'aaa', but got 0138e8ed-87df-4dfe-8b47-4e93ff2ea5bc.
      23:             New-Guid2 | Should be $testCase.String
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\Pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, D:\MockForeachTest.ps1: line 23
    [-] Imported Cmdlet: Test2 123ms
      Expected 'bbb', but got 703a6aab-c24d-419b-9ae9-abb01f017a91.
      23:             New-Guid2 | Should be $testCase.String
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\Pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, D:\MockForeachTest.ps1: line 23
    [-] Imported Cmdlet: Test3 26ms
      Expected 'ccc', but got ae4b1e38-5120-4645-80db-67d6f325d1c8.
      23:             New-Guid2 | Should be $testCase.String
      at Invoke-LegacyAssertion, C:\Program Files\WindowsPowerShell\Modules\Pester\4.4.0\Functions\Assertions\Should.ps1: line 188
      at <ScriptBlock>, D:\MockForeachTest.ps1: line 23
Tests completed in 1.68s
Tests Passed: 3, Failed: 3, Skipped: 0, Pending: 0, Inconclusive: 0 
```

**Comment by mhendric** [comment]
Here's one more piece for the puzzle. It looks like I get different results based on if I execute the test script by hitting the "Run Script" button in PowerShell ISE, vs if I do Invoke-Pester scriptname, or just call the script by name directly. For 'MockTest.ps1', if I execute via "Run Script" in the ISE, one additional test passes (With Test Cases Different Module - Returning Param). For "MockForeachTest.ps1", if I execute via "Run Script", ALL tests pass.

When executing the scripts via straight PowerShell, via PowerShell ISE (but calling the script name directly, not hitting "Run Script"), or by doing Invoke-Pester scriptname in PowerShell or PowerShell ISE, I get the same failures shown in my previous posts.

**Comment by mikeclayton** [comment]
Hi @mhendric .

I was able to reproduce your issue, but I found that if you remove the ```-ModuleName "TestModule"``` from the Mocks, the tests all start working. I think adding the ```-ModuleName``` tells Pester to create the Mock so that calls to that function are mocked only when made from **inside** the specified module.

Since you're calling ```New-Guid2``` from your **MockTests.ps1** and not from other functions defined in **TestModule.psm1** you don't need the ```-ModuleName```.

I don't know if maybe your real code is a bit more complicated than the example, but I hope this helps.

M
